File: internal/importer/normalize.go
package importer

import (
	"fmt"
	"strings"

	"ledger/internal/store"
)

// NormalizedRow is a source row after field resolution. Err is non-nil for rows
// that failed to parse; they are counted as rows_error and skipped by the importer.
type NormalizedRow struct {
	Txn          store.TransactionRow
	CategoryName string // resolved canonical category name; empty when no category column
	RowIndex     int    // 1-based row number (for error messages)
	Err          error
}

// Normalize converts one RawRow to a NormalizedRow using the column/format config.
// The returned Txn always has Source="import" and Status="needs_review"; category
// assignment and status promotion happen in the Importer after a successful insert.
func Normalize(raw RawRow, m MapConfig, rowIndex int) NormalizedRow {
	n := NormalizedRow{RowIndex: rowIndex}

	// Date
	dateStr := strings.TrimSpace(raw[m.Columns.Date])
	if dateStr == "" {
		n.Err = fmt.Errorf("row %d: date column %q is empty", rowIndex, m.Columns.Date)
		return n
	}
	t, err := m.ParseDate(dateStr)
	if err != nil {
		n.Err = fmt.Errorf("row %d: %w", rowIndex, err)
		return n
	}
	n.Txn.PostedAt = t

	// Description / merchant
	desc := strings.TrimSpace(raw[m.Columns.Description])
	if desc == "" {
		n.Err = fmt.Errorf("row %d: description column %q is empty", rowIndex, m.Columns.Description)
		return n
	}
	n.Txn.MerchantRaw = desc

	// Amount + direction
	amountFils, direction, err := m.ParseAmount(raw)
	if err != nil {
		n.Err = fmt.Errorf("row %d: %w", rowIndex, err)
		return n
	}
	n.Txn.AmountFils = amountFils
	n.Txn.Direction = direction

	// Fixed fields
	n.Txn.Currency = m.Currency
	n.Txn.Source = "import"
	n.Txn.Status = "needs_review" // promoted by Importer.Run after category assignment

	// Optional category column
	if m.Columns.Category != "" {
		if catName := strings.TrimSpace(raw[m.Columns.Category]); catName != "" {
			n.CategoryName = m.ResolveCategory(catName)
		}
	}

	return n
}

File: client/src/importer/normalize.ts
import type { RawRow } from "./csv";
import { parseAmount, parseImportDate, validateMap, type ImportMap } from "./map";

export interface NormalizedImportRow { rowIndex: number; postedAt: string; merchantRaw: string; amountMinor: bigint; currency: string; direction: "debit" | "credit"; category: string | null }
export type NormalizeResult = { ok: true; row: NormalizedImportRow } | { ok: false; rowIndex: number; error: string };

export function normalizeRow(raw: RawRow, map: ImportMap, rowIndex: number): NormalizeResult {
  const configErrors = validateMap(map); if (configErrors.length) return { ok: false, rowIndex, error: configErrors.join(" ") };
  try {
    const date = (raw[map.columns.date] ?? "").trim(); if (!date) throw new Error(`date column ${JSON.stringify(map.columns.date)} is empty`);
    const merchantRaw = (raw[map.columns.description] ?? "").trim(); if (!merchantRaw) throw new Error(`description column ${JSON.stringify(map.columns.description)} is empty`);
    const amount = parseAmount(raw, map);
    if (map.skipZeroAmounts && amount.amountMinor === 0n) throw new Error("zero amount skipped");
    const sourceCategory = map.columns.category ? (raw[map.columns.category] ?? "").trim() : "";
    return { ok: true, row: { rowIndex, postedAt: parseImportDate(date, map.dateFormat), merchantRaw, ...amount, currency: map.currency, category: sourceCategory === "" ? null : map.categories?.[sourceCategory] ?? sourceCategory } };
  } catch (error) { return { ok: false, rowIndex, error: error instanceof Error ? error.message : String(error) }; }
}

export function normalizeRows(rows: readonly RawRow[], map: ImportMap): NormalizeResult[] { return rows.map((row, i) => normalizeRow(row, map, i + 1)); }
